x-pack/plugins/security/server/anonymous_access/anonymous_access_service.ts

```diff
@@ -0,0 +1,140 @@
+ */
+
+import type { Request } from '@hapi/hapi';
+import {
+  CapabilitiesStart,
+  IBasePath,
+  KibanaRequest,
+  Logger,
+  Capabilities,
+} from '../../../../../src/core/server';
+import { addSpaceIdToPath } from '../../../spaces/common';
+import type { SpacesServiceStart } from '../../../spaces/server';
+import { AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER } from '../../common/constants';
+import { AnonymousAuthenticationProvider } from '../authentication';
+import type { ConfigType } from '../config';
+import { getDetailedErrorMessage, getErrorStatusCode } from '../errors';
+
+export interface AnonymousAccessServiceStart {
+  readonly isAnonymousAccessEnabled: boolean;
+  // We cannot use `ReadonlyMap` just yet: https://github.com/microsoft/TypeScript/issues/16840
+  readonly accessURLParameters: Readonly<Map<string, string>> | null;
+  getCapabilities: (request: KibanaRequest) => Promise<Capabilities>;
+}
+
+interface AnonymousAccessServiceStartParams {
+  basePath: IBasePath;
+  capabilities: CapabilitiesStart;
+  spaces?: SpacesServiceStart;
+}
+
+const DEFAULT_CAPABILITIES: Capabilities = { navLinks: {}, management: {}, catalogue: {} };
+
+/**
+ * Service that manages various aspects of the anonymous access.
+ */
+export class AnonymousAccessService {
+  /**
+   * Indicates whether anonymous access is enabled.
+   */
+  private isAnonymousAccessEnabled = false;
+
+  constructor(private readonly logger: Logger, private readonly getConfig: () => ConfigType) {}
+
+  setup() {
+    this.isAnonymousAccessEnabled = this.getConfig().authc.sortedProviders.some(
+      ({ type }) => type === AnonymousAuthenticationProvider.type
+    );
+  }
+
+  start({
+    basePath,
+    capabilities,
+    spaces,
+  }: AnonymousAccessServiceStartParams): AnonymousAccessServiceStart {
+    const config = this.getConfig();
+    const anonymousProvider = config.authc.sortedProviders.find(
+      ({ type }) => type === AnonymousAuthenticationProvider.type
+    );
+    // We don't need to add any special parameters to the URL if any of the following is true:
+    // * anonymous provider isn't enabled
+    // * anonymous provider is enabled, but it's a default authentication mechanism
+    const anonymousIsDefault =
+      !config.authc.selector.enabled && anonymousProvider === config.authc.sortedProviders[0];
+
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled,
+      accessURLParameters:
+        anonymousProvider && !anonymousIsDefault
+          ? Object.freeze(
+              new Map([[AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER, anonymousProvider.name]])
+            )
+          : null,
+      getCapabilities: async (request) => {
+        this.logger.debug('Retrieving capabilities of the anonymous service account.');
+
+        if (!this.isAnonymousAccessEnabled) {
+          this.logger.warn('Anonymous access is not enabled');
+          return DEFAULT_CAPABILITIES;
```

Comment:
We recently introduced the concept of "default capabilities" into core, so I think we should instead use that mechanism here. This will allow us to return the full set of UI capabilities, so consumers won't have to worry as much about undefined properties on the capabilities object.

We'll need to update the [`resolveCapabilities`](https://github.com/elastic/kibana/blob/d6cf633e4f0fcc16eb38f9597f08ac95b587a8de/src/core/server/capabilities/capabilities_service.ts#L165) signature to allow for an optional `useDefaultCapabilities` boolean, which we can pass through to the underlying function which currently defaults to `false`.

So the call here would instead look _something_ like:
```suggestion
          return capabilities.resolveCapabilities(fakeRequest, true);
```

Reply:
Oh, interesting, didn't know that. It sounds like a way to go then, but it can be a bit tricky for the case when anonymous service account isn't properly configured and we'll get 401. If I understand everything correctly we have 3 different cases here:

* Anonymous access is not enabled:
```ts
const fakeAnonymousRequest = KibanaRequest.from({ auth: { isAuthenticated: false } });
const caps = await capabilities.resolveCapabilities(fakeAnonymousRequest, { useDefaultCaps: true });
```
* Anonymous access is enabled and properly configured:
```ts
const fakeAnonymousRequest = KibanaRequest.from({ auth: { isAuthenticated: true } });
const caps = await capabilities.resolveCapabilities(fakeAnonymousRequest);
```

* Anonymous access is enabled but misconfigured (we don't know that in advance):
```ts
const fakeAnonymousRequest = KibanaRequest.from({ auth: { isAuthenticated: true } });
const caps = await capabilities.resolveCapabilities(fakeAnonymousRequest);

// if we get 401 then we'll try again this:

const fakeAnonymousRequest = KibanaRequest.from({ auth: { isAuthenticated: false } });
const caps = await capabilities.resolveCapabilities(fakeAnonymousRequest, { useDefaultCaps: true });
```
For the last case we could also make an additional "pre-flight" request to `_authenticate` and depending on the result switch `useDefaultCaps` and `isAuthenticated` on or off. It may be more correct solution, but I'm not sure if it makes sense to optimize for the case that should be very rare at the cost of an additional request for all cases.

What do you think? I'll go ahead without pre-flight request to `_authenticate` for now, but can quickly adjust.

Reply:
I think you summarized the scenarios correctly here. For the misconfigured case, I don't have a strong opinion either way - `_authenticate` does feel more correct, so I'm +1 as long as it's not too difficult to implement.